File: BennuLib/IO/NativeEncoder.vb
Option Infer On

Imports System.IO.Compression
Imports System.IO
Imports PkSprite.Core

Namespace Bennu.IO
    Public MustInherit Class NativeEncoder(Of T)
        Implements IEncoder(Of T)

        Protected MustOverride ReadOnly Property Version As Byte
        Protected MustOverride Sub WriteNativeFormat(obj As T, writer As NativeFormatWriter)
        Protected MustOverride Function GetFileId(obj As T) As String

        Public ReadOnly Compression As CompressionOptions

        Public Sub New()
            Me.New(CompressionOptions.Uncompressed)
        End Sub

        Public Sub New(compressionOptions As CompressionOptions)
            Compression = compressionOptions
        End Sub

        Public Sub Encode(obj As T, output As Stream) Implements IEncoder(Of T).Encode
            Dim nativeStream As Stream

            If Compression = CompressionOptions.Uncompressed Then
                nativeStream = output
            Else
                nativeStream = New GZipStream(output, DirectCast(Compression, CompressionLevel))
            End If
            Using writer As New NativeFormatWriter(output)
                writer.WriteAsciiZ(GetFileId(obj).Substring(0, 3), 3)
                writer.Write(NativeDescriptor)
                writer.Write(Version)
                WriteNativeFormat(obj, writer)
            End Using
        End Sub

        Public Enum CompressionOptions
            Uncompressed = -1
            Fastest = CompressionLevel.Fastest
            Optimal = CompressionLevel.Optimal
            NoCompression = CompressionLevel.NoCompression ' Todo: Is it the same as Uncompressed?
        End Enum
    End Class
End Namespace

File: BennuLib/BennuLib.Core/IO/NativeEncoder.cs
using System.IO.Compression;
using System.IO;

namespace Bennu.IO
{
    public abstract class NativeEncoder<T> : IEncoder<T>
    {

        protected abstract byte GetLastHeaderByte ( T what );

        protected abstract void WriteNativeFormatBody ( T obj, NativeFormatWriter writer );

        protected abstract string GetFileMagic ( T what );

        public CompressionOptions Compression { get; }

        public NativeEncoder () : this ( CompressionOptions.Uncompressed ) { }

        public NativeEncoder ( CompressionOptions compressionOptions )
        {
            Compression = compressionOptions;
        }

        public void Encode ( T what, Stream output )
        {
            Stream nativeStream = null;

            if ( Compression == CompressionOptions.Uncompressed )
            {
                nativeStream = output;
            }
            else
            {
                nativeStream = new GZipStream ( output, ( CompressionLevel ) Compression );
            }

            using ( NativeFormatWriter writer = new NativeFormatWriter ( output ) )
            {
                writer.WriteAsciiZ ( GetFileMagic ( what ).Substring ( 0, 3 ), 3 );
                writer.Write ( NativeFormat.Terminator );
                writer.Write ( GetLastHeaderByte ( what ) );
                WriteNativeFormatBody ( what, writer );
            }
        }

        public enum CompressionOptions
        {
            Uncompressed = -1,
            Fastest = CompressionLevel.Fastest,
            Optimal = CompressionLevel.Optimal,
            NoCompression = CompressionLevel.NoCompression
            // Todo: Is it the same as Uncompressed?
        }
    }
}
